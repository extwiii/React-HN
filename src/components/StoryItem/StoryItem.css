button {
  font-size: 1.3em;
  grid-row: 2;
  grid-column: 2;
  color: #ff6600;
  background-color: white;
  border: solid 1px #ff6600;
  cursor: pointer;
  height: 30px;
  border-radius: 4px;
  transition: 0.2s all ease-in-out;
}

button:hover {
  background: linear-gradient(135deg, #ff6600, red);
  text-shadow: 0 0 1px #fff;
  border: none;
  color: white;
  box-shadow: 6px 8px 10px rgba(0, 0, 0, 0.1);
}

.story_item {
  display: grid;
  grid-template: 90px 1fr / 15% 1fr;
  font-size: 0.6em;
  padding: 0.5em;
  margin: 0.5em;
  background-color: white;
  border-radius: 4px;
  box-shadow: 6px 8px 10px rgba(0, 0, 0, 0.1);
  transition: 0.2s box-shadow ease-in-out;
}

.story_item.open {
  grid-template: 75px 1fr / 1fr 6fr;
}

.story_item:hover {
  box-shadow: 3px 4px 10px rgba(0, 0, 0, 0.1);
}

.story_item_score {
  padding-top: 25%;
  text-align: center;
  grid-row: 1;
  grid-column: 1;
}
.story_item_title {
  grid-row: 1;
  grid-column: 2/3;
}

.story_item_title > a {
  color: black;
  text-decoration: none;
}
.story_item_title > a:hover {
  color: #ff6600;
}

.story_item_title > a:active,
.story_item_title > a:visited {
  color: #999;
}

.story_item_title > p {
  color: #999;
  font-size: 1.1em;
}

.story_item_comments {
  display: grid;
  grid-template: 1fr / 1fr 120px;
  grid-row: 2;
  grid-column: 1/4;
}

.story_item_comments_text {
  grid-row: 1;
  color: white;
  border-radius: 5px;
  background: linear-gradient(135deg, #ff6600, red);
  font-size: 1.3em;
  grid-column: 1/3;
  margin: 1em;
  overflow: scroll;
  height: 300px;
  box-shadow: 4px 4px 10px rgba(0, 0, 0, 0.3) inset;
}

.story_item_comments_text ul {
  padding: 1.8em;
  padding-top: 1em;
}
.story_item_comments_text li {
  padding-bottom: 1em;
}

.story_item_comments_text a {
  color: blue;
}

.story_item_comments_image {
  display: flex;
  grid-row: 1;
  grid-column: 1/3;
  justify-content: center;
  align-items: center;
  height: 300px;
}
