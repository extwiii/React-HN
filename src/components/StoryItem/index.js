// @flow
import React, { Component } from 'react';
import './StoryItem.css';

type Props = {
  story: {
    by: string,
    score: number,
    title: string,
    url: string,
    kids: ?Array<number>,
    kidsItems: ?Array<{
      by: string,
      id: number,
      kids: ?Array<number>,
      parent: number,
      text: string,
      time: number,
      type: string,
    }>,
  },
  handleGetKidsItems(): void,
  className: string,
};

type State = {
  isOpen: boolean,
};

class StoryItem extends Component<Props, State> {
  state = {
    isOpen: false,
  };

  handleOpen = () => {
    this.setState({
      isOpen: true,
    });
    this.props.handleGetKidsItems();
  };
  handleShrink = () => {
    this.setState({
      isOpen: false,
    });
  };

  renderHTML = (rawHTML: ?string) =>
    React.createElement('span', { dangerouslySetInnerHTML: { __html: rawHTML } });

  render() {
    const {
      by, score, title, url, kids, kidsItems,
    } = this.props.story;
    const { className } = this.props;
    const { isOpen } = this.state;

    let items;

    if (kids) {
      if (kidsItems && kids.length === kidsItems.length) {
        items = (
          <div className="story_item_comments">
            <div className="story_item_comments_text">
              <ul>
                {kidsItems.map((item) => {
                  if (item) {
                    if (item.text) {
                      return (
                        <li key={item.id}>{this.renderHTML(item.text)}</li> || <li>Undefined</li>
                      );
                    }
                    return (
                      <li key={item.id} style={{ color: 'black' }}>
                        This entry is deleted
                      </li>
                    );
                  }
                  return false;
                })}
              </ul>
            </div>
            <button onClick={this.handleShrink}>Shrink</button>
          </div>
        );
      } else {
        items = (
          <div className="story_item_comments">
            <div className="story_item_comments_image">
              <img
                src="http://maffrigby.com/wp-content/uploads/2014/03/gettestr-large-spinner-orange.gif"
                alt="hacker news rectangle spinner"
              />
            </div>
            <button onClick={this.handleShrink}>Waiting...</button>
          </div>
        );
      }
    } else {
      items = (
        <div className="story_item_comments">
          <div className="story_item_comments_image">
            <h2>No Comment Yet!</h2>
          </div>
          <button onClick={this.handleShrink}>Waiting...</button>
        </div>
      );
    }
    return (
      <div className={isOpen ? `${className} open` : className}>
        <div className="story_item_score">
          <img src="https://news.ycombinator.com/grayarrow2x.gif" alt="upvote gif" />
          <p>{score}</p>
        </div>
        <div className="story_item_title">
          <a href={url} target="_blank">
            <h2>{title}</h2>
          </a>
          <p>
            by <strong>{by}</strong>
          </p>
        </div>
        {isOpen ? (
          items
        ) : (
          <div className="story_item_comments">
            <button onClick={this.handleOpen}>Read More</button>
          </div>
        )}
      </div>
    );
  }
}

export default StoryItem;
