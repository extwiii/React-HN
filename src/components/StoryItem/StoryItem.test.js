import React from 'react';
import { shallow } from 'enzyme';

import StoryItem from './index';

describe('StoryItem', () => {
  let props;
  let mountedStoryItem;
  const storyItem = () => {
    if (!mountedStoryItem) {
      mountedStoryItem = shallow(<StoryItem {...props} />);
    }
    return mountedStoryItem;
  };

  test('renders props correctly ', () => {
    expect(storyItem().instance().props).toEqual(props);
    expect(storyItem().props().className).toEqual(props.className);
  });

  test('renders correctly if isOpen is falsy', () => {
    expect(storyItem().instance().state.isOpen).toEqual(false);
    expect(storyItem().find('.story_item_comments').length).toEqual(1);
  });

  test('handleOpen function works correctly ', () => {
    expect(storyItem().instance().state.isOpen).toEqual(false);
    storyItem()
      .instance()
      .handleOpen();
    expect(storyItem().instance().state.isOpen).toEqual(true);
  });

  test('handleShrink function works correctly ', () => {
    expect(storyItem().instance().state.isOpen).toEqual(false);
    storyItem()
      .instance()
      .handleOpen();
    expect(storyItem().instance().state.isOpen).toEqual(true);
    storyItem()
      .instance()
      .handleShrink();
    expect(storyItem().instance().state.isOpen).toEqual(false);
  });

  test('renderHTML function works correctly ', () => {
    const renderedHTML = storyItem()
      .instance()
      .renderHTML('First &middot; Second');
    // must check later
    expect(renderedHTML).toEqual(<span dangerouslySetInnerHTML={{ __html: 'First &middot; Second' }} />);
  });

  test('render state correctly', () => {
    expect(storyItem().instance().state).toEqual({ isOpen: false });
  });
  test('render items correctly if there is not kids', () => {
    delete props.story.kids;
    const mountedStoryItemwithNewProps = shallow(<StoryItem {...props} />);
    expect(mountedStoryItemwithNewProps.instance().props).toEqual(props);
  });

  test('render items correctly if there is kids', () => {
    const kids = [123];
    props = {
      ...props,
      story: { ...props.story, kids },
    };
    const mountedStoryItemwithNewProps = shallow(<StoryItem {...props} />);
    expect(mountedStoryItemwithNewProps.instance().props).toEqual(props);
  });

  test('render items correctly if kids and kidsItems have same length', () => {
    const kidsItems = { id: 123, text: 'this is kidsItems text' };
    const kids = [123];
    props = {
      ...props,
      story: { ...props.story, kids, kidsItems },
    };
    const mountedStoryItemwithNewProps = shallow(<StoryItem {...props} />);
    expect(mountedStoryItemwithNewProps.instance().props).toEqual(props);
  });

  test('maps items correctly', () => {
    const kidsItems = [
      {
        id: 123,
        text: 'this is kidsItems text for 123',
      },
      {
        id: 124,
        text: 'this is kidsItems text for 124',
      },
      {
        id: 125,
        text: 'this is kidsItems text for 125',
      },
    ];
    const kids = [123, 124, 125];
    props = {
      ...props,
      story: { ...props.story, kids, kidsItems },
    };
    const mountedStoryItemwithNewProps = shallow(<StoryItem {...props} />);
    mountedStoryItemwithNewProps.instance().handleOpen();
    expect(kidsItems[0].text).toEqual('this is kidsItems text for 123');
  });

  test('maps items if there is no item correctly', () => {
    const kidsItems = [];
    const kids = [123, 124, 125];
    props = {
      ...props,
      story: { ...props.story, kids, kidsItems },
    };
    const mountedStoryItemwithNewProps = shallow(<StoryItem {...props} />);
    mountedStoryItemwithNewProps.instance().handleOpen();
    expect(kidsItems).toEqual([]);
  });

  test('maps items if there is item but no item text correctly', () => {
    const kidsItems = [
      {
        id: 123,
        text: '',
      },
      undefined,
      {
        id: 125,
        text: 'this is kidsItems text for 125',
      },
    ];
    const kids = [123, 124, 125];
    props = {
      ...props,
      story: { ...props.story, kids, kidsItems },
    };
    const mountedStoryItemwithNewProps = shallow(<StoryItem {...props} />);
    mountedStoryItemwithNewProps.instance().handleOpen();
    expect(kidsItems[0].text).toEqual('');
  });

  test('renders DOM correctly', () => {
    const divs = storyItem().find('div');
    const h3s = storyItem().find('h3');
    expect(divs.length).toBeGreaterThan(0);
    expect(divs.length).toBe(4);
    expect(h3s.length).toBe(0);
  });

  test('to match snapshot', () => {
    expect(storyItem()).toMatchSnapshot();
  });

  beforeEach(() => {
    props = {
      story: {
        by: 'by',
        title: 'title',
        kids: [],
        kidsItems: [],
        text: 'text',
        url: 'url',
      },
      className: 'story_item',
      handleGetKidsItems: jest.fn(),
    };
    mountedStoryItem = undefined;
  });
});
