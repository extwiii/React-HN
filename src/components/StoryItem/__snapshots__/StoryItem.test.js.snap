// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StoryItem to match snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <StoryItem
    className="story_item"
    handleGetKidsItems={[Function]}
    story={
        Object {
            "by": "by",
            "kids": Array [],
            "kidsItems": Array [],
            "text": "text",
            "title": "title",
            "url": "url",
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="story_item_score"
>
          <img
                    alt="upvote gif"
                    src="https://news.ycombinator.com/grayarrow2x.gif"
          />
          <p />
</div>,
        <div
          className="story_item_title"
>
          <a
                    href="url"
                    target="_blank"
          >
                    <h2>
                              title
                    </h2>
          </a>
          <p>
                    by 
                    <strong>
                              by
                    </strong>
          </p>
</div>,
        <div
          className="story_item_comments"
>
          <button
                    onClick={[Function]}
          >
                    Read More
          </button>
</div>,
      ],
      "className": "story_item",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <img
              alt="upvote gif"
              src="https://news.ycombinator.com/grayarrow2x.gif"
/>,
            <p />,
          ],
          "className": "story_item_score",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "alt": "upvote gif",
              "src": "https://news.ycombinator.com/grayarrow2x.gif",
            },
            "ref": null,
            "rendered": null,
            "type": "img",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": undefined,
            },
            "ref": null,
            "rendered": null,
            "type": "p",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <a
              href="url"
              target="_blank"
>
              <h2>
                            title
              </h2>
</a>,
            <p>
              by 
              <strong>
                            by
              </strong>
</p>,
          ],
          "className": "story_item_title",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <h2>
                title
</h2>,
              "href": "url",
              "target": "_blank",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "title",
              },
              "ref": null,
              "rendered": "title",
              "type": "h2",
            },
            "type": "a",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "by ",
                <strong>
                  by
</strong>,
              ],
            },
            "ref": null,
            "rendered": Array [
              "by ",
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "by",
                },
                "ref": null,
                "rendered": "by",
                "type": "strong",
              },
            ],
            "type": "p",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <button
            onClick={[Function]}
>
            Read More
</button>,
          "className": "story_item_comments",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Read More",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Read More",
          "type": "button",
        },
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="story_item_score"
>
            <img
                        alt="upvote gif"
                        src="https://news.ycombinator.com/grayarrow2x.gif"
            />
            <p />
</div>,
          <div
            className="story_item_title"
>
            <a
                        href="url"
                        target="_blank"
            >
                        <h2>
                                    title
                        </h2>
            </a>
            <p>
                        by 
                        <strong>
                                    by
                        </strong>
            </p>
</div>,
          <div
            className="story_item_comments"
>
            <button
                        onClick={[Function]}
            >
                        Read More
            </button>
</div>,
        ],
        "className": "story_item",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <img
                alt="upvote gif"
                src="https://news.ycombinator.com/grayarrow2x.gif"
/>,
              <p />,
            ],
            "className": "story_item_score",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "alt": "upvote gif",
                "src": "https://news.ycombinator.com/grayarrow2x.gif",
              },
              "ref": null,
              "rendered": null,
              "type": "img",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": undefined,
              },
              "ref": null,
              "rendered": null,
              "type": "p",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <a
                href="url"
                target="_blank"
>
                <h2>
                                title
                </h2>
</a>,
              <p>
                by 
                <strong>
                                by
                </strong>
</p>,
            ],
            "className": "story_item_title",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <h2>
                  title
</h2>,
                "href": "url",
                "target": "_blank",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "title",
                },
                "ref": null,
                "rendered": "title",
                "type": "h2",
              },
              "type": "a",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "by ",
                  <strong>
                    by
</strong>,
                ],
              },
              "ref": null,
              "rendered": Array [
                "by ",
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "by",
                  },
                  "ref": null,
                  "rendered": "by",
                  "type": "strong",
                },
              ],
              "type": "p",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <button
              onClick={[Function]}
>
              Read More
</button>,
            "className": "story_item_comments",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Read More",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Read More",
            "type": "button",
          },
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
